### ISSUE TITLE
Large output data handling

### ISSUE THREAD

**Issue by znicholls** [openclimatedata/openscm#54]
**Is your feature request related to a problem? Please describe.**

When running an ensemble of parameter sets for multiple scenarios, the outputs can quickly become rather large and hard to handle.

**Describe the solution you'd like**

Some way for us to handle these larger data sets. Speaking with @chrisroadmap, a nice way to save the data would be into netCDFs where the outputs are output data variables (each of which is at least 2D, one axis is run number, other is time, other axes for regions etc. as necessary) or parameters (1D, the axis being run number). The information about how the file was saved can be saved into the attributes of the netCDF. We would then need to provide functions to write such netCDF files, read them as well as run just the parameter sets (for reproducibility/to give people a standard parameter set format they can use to run models). Handling memory usage is another issue, I'm not sure if Pandas does anything smart or whether we'd have to switch to e.g. xarray for this...

Using netCDF will require rethinking how we install. Here it is probably best if a netCDF library is optional and if not available, OpenSCM tries to save everything as csvs. This means that it is possible to do large ensembles but people who just want to fiddle around don't have to muck around with netCDF libraries.

**Describe alternatives you've considered**

Use csvs only with some custom way of saving files into suitably small chunks. Sounds like reinventing the wheel given the efforts that have gone into big data handling with netCDF...

**Additional context**

**Comment by swillner** [comment]
Maybe we should define several output formats (#11), for single runs (CSV) and ensembles (NetCDF)?

**Comment by znicholls** [comment]
> Maybe we should define several output formats (#106), for single runs (CSV) and ensembles (NetCDF)?

Yep although output format might depend on what is available rather than number of runs (with some sensible warning like if it's more than 10 runs and netCDF isn't available then you get a big warning saying csv is likely to explode).

**Comment by znicholls** [comment]
For our netcdf format I think we could do something like:

two dimensions:

1. time
1. run_id (i.e. a unique identifier of each model run)

scalar parameters are stored as 1D variables, where each parameter value corresponds to a specific run_id
timeseries are stored as 2D variables, each timeseries has time as one dimension and run_id as the other dimension

Other metadata can be stored in the attributes of the netcdf file.

The thing I'm still not sure about is how to handle generic parameters. You can't just stick anything in a netCDF variable so we wouldn't be able to handle some of the stranger use cases (e.g. generic parameters which are themselves lists). Perhaps we need to write a custom way to stringify and then change back to handle this particular case? Having said that, as a starting point I'd suggest we try to save generic parameters like scalar parameters and just throw errors if the generic type is incompatible (the errors should point users to raise an issue in OpenSCM so we have a clear use case if/when we come to implementing something custom).

Given our data is going to be custom, I don't think it will play nice with iris as iris is so opinionated about things. I suspect [xarray](https://anaconda.org/conda-forge/xarray) will be our best bet and we'll just to write an extra layer to handle generic parameters when the time comes.
